sdk/src/derivation_path.rs

```diff
@@ -1,94 +1,90 @@
+#[derive(PartialEq)]
+pub struct DerivationPath(DerivationPathInner);
 
-    pub fn as_u32(&self) -> u32 {
-        self.0
+impl Default for DerivationPath {
+    fn default() -> Self {
+        Self::new_bip44(Solana, None, None)
     }
 }
 
-impl From<u32> for DerivationPathComponent {
-    fn from(n: u32) -> Self {
-        Self(n | Self::HARDENED_BIT)
+impl DerivationPath {
+    fn new<P: Into<Box<[ChildIndex]>>>(path: P) -> Self {
+        Self(DerivationPathInner::new(path))
     }
-}
 
-impl FromStr for DerivationPathComponent {
-    type Err = DerivationPathError;
+    pub fn from_key_str<T: Bip44>(path: &str, coin: T) -> Result<Self, DerivationPathError> {
+        let path = format!("m/{}", path);
+        let extend = DerivationPathInner::from_str(&path)
+            .map_err(|err| DerivationPathError::InvalidDerivationPath(err.to_string()))?;
+        let mut extend = extend.into_iter();
+        let account = extend.next().map(|index| index.to_u32());
+        let change = extend.next().map(|index| index.to_u32());
+        if extend.next().is_some() {
+            return Err(DerivationPathError::InvalidDerivationPath(format!(
+                "key path `{}` too deep, only <account>/<change> supported",
+                path
+            )));
+        }
+        Ok(Self::new_bip44(coin, account, change))
+    }
 
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let index_str = if let Some(stripped) = s.strip_suffix('\'') {
-            stripped
-        } else {
-            s
-        };
-        index_str.parse::<u32>().map(|ki| ki.into()).map_err(|_| {
-            DerivationPathError::InvalidDerivationPath(format!(
-                "failed to parse path component: {:?}",
-                s
-            ))
-        })
+    fn _from_absolute_path_str(path: &str) -> Result<Self, DerivationPathError> {
+        let inner = DerivationPath::_from_absolute_path_insecure_str(path)?
+            .into_iter()
+            .map(|c| ChildIndex::Hardened(c.to_u32()))
+            .collect::<Vec<_>>();
+        Ok(Self(DerivationPathInner::new(inner)))
     }
-}
 
-impl std::fmt::Display for DerivationPathComponent {
-    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-        let hardened = if (self.0 & Self::HARDENED_BIT) == 0 {
-            ""
-        } else {
-            "'"
-        };
-        let index = self.0 & !Self::HARDENED_BIT;
-        write!(fmt, "{}{}", index, hardened)
+    fn _from_absolute_path_insecure_str(path: &str) -> Result<Self, DerivationPathError> {
+        Ok(Self(DerivationPathInner::from_str(&path).map_err(
+            |err| DerivationPathError::InvalidDerivationPath(err.to_string()),
+        )?))
     }
-}
 
-impl std::fmt::Debug for DerivationPathComponent {
-    fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
-        std::fmt::Display::fmt(self, fmt)
+    pub fn new_bip44<T: Bip44>(coin: T, account: Option<u32>, change: Option<u32>) -> Self {
```

Comment:
Let's rename this to `new_bip44_with_coin`, add a pub wrapper that calls with `Solana` and make all of the trait referencing bits private for now.  I'm tempted to say our API referencing BIP44 can assume the Solana address and any other implementer can take the `full-path`/BIP32 route.  But I'm no so confident as to reject the trait at this point